### ISSUE TITLE
imap4 LOGINCredentials is not generic for smtp use

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#4692]
|<img alt="marcuslow's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">| marcuslow reported|
|-|-|
|Trac ID|trac#4692|
|Type|enhancement|
|Created|2010-10-15 15:50:59Z|
|Branch|https://github.com/twisted/twisted/tree/smtp-login-4692|

twisted mail smtp does not have `LOGINCredentials` so registering `imap4.LOGINCredentials` is expected to work, but it doesn't work with popular email clients in the real world.

The issue is the `self.challenge` initialized at `__init__` uses `"User Name\0"` whereelse clients like Outlook/OE checks for `"Username:"`.

It would be easy though to create the same class in smtp and just override the `"User Name\0"` and `"Password\0"` with `"Username:"` and `"Password:"`

Attachments:

* [4692.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/07e/07e95121f318dc32d21947cc8e08a03b63ce155e/c3fb698f68ec71f1d2645e4b26a558b16202dc12.patch) (552 bytes) - added by retenodus on 2011-06-30 19:35:04Z - 
* [4692-smtp.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/07e/07e95121f318dc32d21947cc8e08a03b63ce155e/03f37c6da7bdfd2565a8b2a1da97fcfe8c9e0909.patch) (492 bytes) - added by retenodus on 2011-06-30 23:14:45Z - 
* [4692-test_smtp.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/07e/07e95121f318dc32d21947cc8e08a03b63ce155e/1a1d5abe8166117e2063cda54f0b0574b5491710.patch) (603 bytes) - added by retenodus on 2011-06-30 23:14:59Z - 

<details><summary>Searchable metadata</summary>

```
trac-id__4692 4692
type__enhancement enhancement
reporter__marcuslow marcuslow
priority__normal normal
milestone__ 
branch__branches_smtp_login_4692 branches/smtp-login-4692
branch_author__retenodus retenodus
status__closed closed
resolution__fixed fixed
component__mail mail
keywords__ 
time__1287157859000000 1287157859000000
changetime__1312587698000000 1312587698000000
version__None None
owner__exarkun exarkun
cc__retenodus
```
</details>

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|

**Comment by twisted-trac** [comment]
|<img alt="retenodus's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_4"></a>|retenodus commented|
|-|-|

The class is created, is it good ?

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_5"></a>|@exarkun set owner to retenodus|
|-|-|

Thanks for the patch retenodus.  This looks like a step in the right direction.  The feature needs to be unit tested and the new class needs a docstring as well.

I'm looking forward to your next patch. :)

**Comment by twisted-trac** [comment]
|<img alt="retenodus's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_6"></a>|retenodus removed owner|
|-|-|

I've added 2 files : one is the unittest, where I've juste deleted the WeirdLogin class (not the "normal" login class which still uses imap4.LOGINCredentials).

The other is the new version of smtp.py, with the docstrings.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_7"></a>|@exarkun commented|
|-|-|

Thanks!  Some more things:

  1. It's usually a good idea to just generate one patch file.  It's less work getting one patch attached to the ticket than three, and less work applying them too. :)
  1. Generate the new patch against trunk, so it includes all of the changes, rather than against the previously submitted patch.
  1. Generate patches from the root of a trunk checkout, so that they have useful path names and can be applied without knowing which directory the files they modify are present in.

Basically, cd into your trunk checkout and just do "svn diff > some.patch" and you'll get roughly the right thing.

I'm going to check the patches in to a branch.  My main other concern is verifying that SMTP clients actually work this way.  The draft specification linked from the unit tests, <http://sepp.oetiker.ch/sasl-2.1.19-ds/draft-murchison-sasl-login-00.txt>, suggests that 'User Name\x00' is the correct challenge.  I don't have Outlook available for testing.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_8"></a>|@exarkun commented|
|-|-|

(In [32206]) Branching to 'smtp-login-4692'

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_9"></a>|@exarkun commented|
|-|-|

http://www.melnikov.ca/mel/devel/Links.html also seems a helpful reference.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_10"></a>|@exarkun commented|
|-|-|

(In [32207]) Apply 4692.patch, 4692-smtp.patch, 4692-test_smtp.patch

refs [#4692](https://github.com/twisted/twisted/issues/4692)

**Comment by twisted-trac** [comment]
|<img alt="retenodus's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_12"></a>|retenodus commented|
|-|-|

Hi,

Thanks for your advices.
Euh, if I understand my work is over for this ticket, right ? I'm not sure that I've correctly understood you.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_13"></a>|@exarkun commented|
|-|-|

Yes, at least until someone does another review. :)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_14"></a>|@exarkun set owner to @exarkun|
|-|-|

Okay, I tested with some clients.  I did reproduce the problem on Outlook with the old code, and with the new code things work better.  I also tested Thunderbird, which appears to be happy either way.  So I guess this is the right thing to do.  The only changes I think should be made are:

  1. [ReviewProcess#Newsfiles Add a news file]
  1. Keep `lcredentials` out of the public namespace - use `_lcredentials`
  1. The `smtp.LOGINCredentials` docstring isn't particularly clear; it doesn't really matter what it subclasses, and I'm not sure what "set credentials to authenticate" means.  It should probably also mention the reason it is contravening the specification.

These are trivial, I'll make the changes and then merge.  Thanks!

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_15"></a>|@exarkun set status to `closed`|
|-|-|

(In [32426]) Merge smtp-login-4692 

Author: retenodus
Reviewer: exarkun
Fixes: [#4692](https://github.com/twisted/twisted/issues/4692)

Change `twisted.mail.smtp.LOGINCredentials` to emit Outlook-compatible
challenge strings.
